crates/tui/src/config_tab.rs

```diff
@@ -0,0 +1,311 @@
+use std::sync::Mutex;
+
+use alloy::primitives::Address;
+use alloy_chains::NamedChain;
+use crossterm::event::{KeyCode, KeyEvent};
+use ratatui::{
+    layout::{Constraint, Layout},
+    prelude::{Buffer, Rect},
+    style::{Style, Stylize},
+    widgets::{Block, List, ListState, Padding, Row, StatefulWidget, Table, Widget},
+};
+
+use crate::{config::Config, HandleEvent};
+
+pub struct ConfigTab {
+    config: Config,
+    config_list_state: Mutex<ListState>,
+    origin_connections_collapsed: bool,
+    labels_collapsed: bool,
+}
+
+#[derive(Debug)]
+enum ConfigListItemType {
+    Rpcs,
+    OriginConnections(Address),
+    OriginConnectionsMeta,
+    Labels(NamedChain),
+    LabelsMeta,
+    Meta,
+}
+
+impl ConfigTab {
+    pub fn new(config: Config) -> Self {
+        let mut list_state = ListState::default();
+        list_state.select_first();
+
+        Self {
+            config,
+            config_list_state: Mutex::new(list_state.clone()),
+            origin_connections_collapsed: false,
+            labels_collapsed: false,
+        }
+    }
+
+    fn list_len(&self) -> usize {
+        3 + if self.origin_connections_collapsed {
+            0
+        } else {
+            self.config.origin_connections.len()
+        } + if self.labels_collapsed {
+            0
+        } else {
+            self.config.labels.len()
+        }
+    }
+
+    fn origin_connections_offset(&self) -> usize {
+        1
+    }
+
+    fn labels_offset(&self) -> usize {
+        2 + if self.origin_connections_collapsed {
+            0
+        } else {
+            self.config.origin_connections.len()
+        }
+    }
+
+    /// Sorted keys to get consistent order since order of keys in HashMap::keys is unknown.
+    fn origin_connections_keys(&self) -> Vec<&Address> {
+        let mut keys = self.config.origin_connections.keys().collect::<Vec<_>>();
+        keys.sort();
+        keys
+    }
+
+    /// Sorted keys to get consistent order since order of keys in HashMap::keys is unknown.
+    fn labels_keys(&self) -> Vec<&NamedChain> {
+        let mut keys = self.config.labels.keys().collect::<Vec<_>>();
+        keys.sort();
+        keys
+    }
+
+    fn nth_addr_origin_connections(&self, idx: usize) -> Option<Address> {
+        self.origin_connections_keys().get(idx).map(|addr| **addr)
+    }
+
+    fn nth_chain_labels(&self, idx: usize) -> Option<NamedChain> {
+        self.labels_keys().get(idx).map(|chain| **chain)
+    }
```

Comment:
These functions are only subsequently used once. Is this early over-optimisation, or do you plan to use this later in other parts of code?